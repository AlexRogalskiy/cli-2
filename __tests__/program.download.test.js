const os = require('os');
const path = require('path');
const git = require('isomorphic-git');
const fsp = require('fs/promises');
const fse = require('fs-extra');
const nock = require('nock');

const programCmd = require('../src/commands/program/download.js');
const initSettings = require('../src/settings.js');
const { readDirP, getFixturePath } = require('./helpers/index.js');

const getTmpDirPath = (program) => path.join(os.tmpdir(), `${program}-program`);

nock.disableNetConnect();

let defaults;

describe('program', () => {
  beforeEach(async () => {
    const tmpDir = await fsp.mkdtemp(path.join(os.tmpdir(), 'hexlet-cli-'));
    defaults = { homedir: tmpDir };
  });

  it('download', async () => {
    const { hexletDir, hexletConfigPath } = initSettings(defaults);
    await fsp.mkdir(hexletDir);
    await fse.writeJson(hexletConfigPath, {});

    const programArchivePath = getFixturePath('ruby-program.tar.gz');
    nock('https://hexlet-programs.fra1.digitaloceanspaces.com')
      .get('/ruby-program.tar.gz')
      .replyWithFile(200, programArchivePath);

    git.clone = jest.fn(() => {});

    const args = {
      program: 'ruby',
      exercise: 'fundamentals',
      token: 'some-token',
      customSettings: defaults,
    };
    await programCmd.handler(args, defaults);

    const tmpDirPath = getTmpDirPath(args.program);
    expect(await readDirP(tmpDirPath)).toMatchSnapshot();

    expect(await readDirP(hexletDir)).toMatchSnapshot();
  });

  it('download (without init)', async () => {
    const args = {
      program: 'ruby',
      exercise: 'hello-world',
      token: 'some-token',
      customSettings: defaults,
    };

    await expect(() => programCmd.handler(args, defaults)).rejects.toThrow('no such file or directory');
  });
});
