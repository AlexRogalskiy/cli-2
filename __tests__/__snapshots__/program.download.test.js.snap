// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`program download 1`] = `
Array [
  Object {
    "name": ".gitignore",
    "path": ".gitignore",
    "size": 20,
    "type": "file",
  },
  Object {
    "name": "TUTORIAL.md",
    "path": "TUTORIAL.md",
    "size": 120,
    "type": "file",
  },
  Object {
    "name": "spec.yml",
    "path": "spec.yml",
    "size": 74,
    "type": "file",
  },
  Object {
    "name": "Makefile",
    "path": "exercises/start/Makefile",
    "size": 23,
    "type": "file",
  },
  Object {
    "name": "README.md",
    "path": "exercises/start/README.md",
    "size": 402,
    "type": "file",
  },
  Object {
    "name": "Gemfile",
    "path": "exercises/fundamentals/Gemfile",
    "size": 70,
    "type": "file",
  },
  Object {
    "name": "Makefile",
    "path": "exercises/fundamentals/Makefile",
    "size": 31,
    "type": "file",
  },
  Object {
    "name": "README.md",
    "path": "exercises/fundamentals/README.md",
    "size": 185,
    "type": "file",
  },
  Object {
    "name": "Rakefile",
    "path": "exercises/fundamentals/Rakefile",
    "size": 205,
    "type": "file",
  },
  Object {
    "name": "fundamentals.gemspec",
    "path": "exercises/fundamentals/fundamentals.gemspec",
    "size": 217,
    "type": "file",
  },
  Object {
    "name": "reverse_test.rb",
    "path": "exercises/fundamentals/test/reverse_test.rb",
    "size": 214,
    "type": "file",
  },
  Object {
    "name": "sum_of_primes_test.rb",
    "path": "exercises/fundamentals/test/sum_of_primes_test.rb",
    "size": 284,
    "type": "file",
  },
  Object {
    "name": "test_helper.rb",
    "path": "exercises/fundamentals/test/test_helper.rb",
    "size": 58,
    "type": "file",
  },
  Object {
    "name": "reverse.rb",
    "path": "exercises/fundamentals/lib/reverse.rb",
    "size": 92,
    "type": "file",
  },
  Object {
    "name": "sum_of_primes.rb",
    "path": "exercises/fundamentals/lib/sum_of_primes.rb",
    "size": 96,
    "type": "file",
  },
  Object {
    "name": "Gemfile",
    "path": "exercises/blocks/Gemfile",
    "size": 70,
    "type": "file",
  },
  Object {
    "name": "Makefile",
    "path": "exercises/blocks/Makefile",
    "size": 57,
    "type": "file",
  },
  Object {
    "name": "README.md",
    "path": "exercises/blocks/README.md",
    "size": 92,
    "type": "file",
  },
  Object {
    "name": "Rakefile",
    "path": "exercises/blocks/Rakefile",
    "size": 205,
    "type": "file",
  },
  Object {
    "name": "blocks.gemspec",
    "path": "exercises/blocks/blocks.gemspec",
    "size": 211,
    "type": "file",
  },
  Object {
    "name": "test_helper.rb",
    "path": "exercises/blocks/test/test_helper.rb",
    "size": 58,
    "type": "file",
  },
  Object {
    "name": "with_numbers_test.rb",
    "path": "exercises/blocks/test/with_numbers_test.rb",
    "size": 322,
    "type": "file",
  },
  Object {
    "name": "with_numbers.rb",
    "path": "exercises/blocks/lib/with_numbers.rb",
    "size": 46,
    "type": "file",
  },
]
`;

exports[`program download 2`] = `
Array [
  Object {
    "name": ".config.json",
    "path": ".config.json",
    "size": 3,
    "type": "file",
  },
  Object {
    "name": ".gitignore",
    "path": "ruby/.gitignore",
    "size": 20,
    "type": "file",
  },
  Object {
    "name": "Gemfile",
    "path": "ruby/exercises/fundamentals/Gemfile",
    "size": 70,
    "type": "file",
  },
  Object {
    "name": "Makefile",
    "path": "ruby/exercises/fundamentals/Makefile",
    "size": 31,
    "type": "file",
  },
  Object {
    "name": "README.md",
    "path": "ruby/exercises/fundamentals/README.md",
    "size": 185,
    "type": "file",
  },
  Object {
    "name": "Rakefile",
    "path": "ruby/exercises/fundamentals/Rakefile",
    "size": 205,
    "type": "file",
  },
  Object {
    "name": "fundamentals.gemspec",
    "path": "ruby/exercises/fundamentals/fundamentals.gemspec",
    "size": 217,
    "type": "file",
  },
  Object {
    "name": "reverse_test.rb",
    "path": "ruby/exercises/fundamentals/test/reverse_test.rb",
    "size": 214,
    "type": "file",
  },
  Object {
    "name": "sum_of_primes_test.rb",
    "path": "ruby/exercises/fundamentals/test/sum_of_primes_test.rb",
    "size": 284,
    "type": "file",
  },
  Object {
    "name": "test_helper.rb",
    "path": "ruby/exercises/fundamentals/test/test_helper.rb",
    "size": 58,
    "type": "file",
  },
  Object {
    "name": "reverse.rb",
    "path": "ruby/exercises/fundamentals/lib/reverse.rb",
    "size": 92,
    "type": "file",
  },
  Object {
    "name": "sum_of_primes.rb",
    "path": "ruby/exercises/fundamentals/lib/sum_of_primes.rb",
    "size": 96,
    "type": "file",
  },
]
`;
